src/topp/SageAdapter.cpp

```diff
@@ -90,6 +96,598 @@ class TOPPSageAdapter :
   {
   }
 
+
+  struct RateMassPair
+{
+  double rate = -1.;
+  double deltamass = -1.;
+};
+
+struct modification{
+  double rate = 0; 
+  vector<double> mass; 
+  double numcharges = 0; 
+}; 
+
+#include <map>
+#include <vector>
+#include <algorithm>
+#include <cmath>
+#include <numeric>
+
+// Gaussian function
+static double gaussian(double x, double sigma) {
+    return exp(-(x*x) / (2 * sigma*sigma)) / (sigma * sqrt(2 * M_PI));
+}
+
+static std::map<double, double> smoothMassSpectrum(const std::map<double, double>& spectrum, double sigma = 0.1) {
```

Comment:
can you add some documentation why smoothed spectra are needed?
or is this on the PTM rate+frequencies?

Reply:
This is on the PTM rate + frequencies, the smoothing is before picking the local maxima in the PTM rate + frequencies histogram. Should I still add the documentation?

Reply:
yes definitely.
and it needs to be clear that this not not about smoothing a spectrum but the histogram(is it a kernel density estimation?)

Reply:
Yes, exactly it's a KDE using a Gaussian. It's not technically absolutely necessary anymore since I moved away from using spline to pick the local maxima, but it also removes some weird, low-rate random masses so I kept it in.